app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,31 @@
+class OmniauthCallbacksController < ApplicationController
+  def complete
+    auth = request.env['omniauth.auth']
+    email_address = auth.info.email
+    govuk_one_login_uid = auth.uid
+
+    candidate = Candidate.find_by(govuk_one_login_uid: govuk_one_login_uid) || Candidate.find_by(email_address: email_address)
+
+    if candidate
+      candidate.update!(govuk_one_login_uid: govuk_one_login_uid) if candidate.govuk_one_login_uid.nil?
+    else
+      candidate = Candidate.create!(email_address: email_address, govuk_one_login_uid: govuk_one_login_uid)
+    end
+
+    session[:onelogin_id_token] = auth.credentials.id_token
+    candidate.update!(last_signed_in_at: Time.zone.now) if sign_in(candidate, scope: :candidate)
+
+    redirect_to root_path
```

Comment:
Presumably we'd check for a redirect path. e.g. Candidate tries to access a part of their form without being signed in, we'd want to redirect them there rather than back to the root